### PR TITLE
CUMULUS-2854: Update queue-granules to correctly set granule createdAt stamp

### PR DESCRIPTION
**Pull request by Jkovarik**
**Summary:** Summary of changes

- Updates queue-granules to set createdAt to the same value as the
Queue write update to avoid granule writes failing in child workflows
- Updates queue-granules queued status update to respect granule
created at value for queued update if the granule timestamp already
exists
- Add unit tests to validate the above
- Adds integration tests to validate the above

Addresses [CUMULUS-2854: Develop amazing new feature](https://bugs.earthdata.nasa.gov/browse/CUMULUS-2854)

## Changes

- **CUMULUS-2854**
  - Fixed queue granules behavior where the task was not accounting for granules that
  *already* had createdAt set. Workflows downstream in this scenario should no longer
  fail to write their granules due to order-of-db-writes constraints in the database
  update logic.

## PR Checklist

- [x] Update CHANGELOG
- [x] Unit tests
- [x] Ad-hoc testing - Deploy changes and test manually
- [x] Integration tests

### CODE REVIEW

File: example/cumulus-tf/queue_granules_passthrough_workflow.tf
Comment (Jkovarik) [inline example/cumulus-tf/queue_granules_passthrough_workflow.tf:15]: This variable/setting for queue granules differs from our other integration tests, as such I duped the workflow.    I'm open to suggestions on time-valuable ways to improve that 🤔

File: tasks/queue-granules/index.js
Comment (Jkovarik) [inline tasks/queue-granules/index.js:51]: This method is the 'guts' of the needed logic change.

File: tasks/queue-granules/tests/index.js
Comment (Jkovarik) [inline tasks/queue-granules/tests/index.js:784]: I'm pretty sure this test wasn't working.   In either case, the premise was incorrect as the record validation is >=

File: tasks/queue-granules/tests/index.js
Comment (Jkovarik) [inline tasks/queue-granules/tests/index.js:812]: Hmm, this isn't a terribly good descriptor 😅

File: tasks/queue-granules/tests/index.js
Comment (Jkovarik) [inline tasks/queue-granules/tests/index.js:387]:
```suggestion
test.serial('The correct message is enqueued with and timestamps are added to granules that are missing them', async (t) => {
```

File: example/spec/parallel/queueGranules/QueueGranulesSpecPostProcessing.input.payload.json
Comment (flamingbear) [inline example/spec/parallel/queueGranules/QueueGranulesSpecPostProcessing.input.payload.json:9]: since these are actually being created, it should be able to run on anyone's stack and these buckets should probably be templated to the user's system bucket.

Comment (flamingbear) [reply example/spec/parallel/queueGranules/QueueGranulesSpecPostProcessing.input.payload.json:9]: This hard coded bucket causes problems with the afterAll cleanup  via the waitForGranuleAndDelete.

**Review by flamingbear — CHANGES_REQUESTED** [review]
A few things that need attention, but overall looks good.  Some are just suggestions, happy to talk through them and what I was thinking if they don't make sense.

File: example/spec/parallel/queueGranules/QueueGranulesSpecPostProcessing.input.payload.json
Comment (Jkovarik) [reply example/spec/parallel/queueGranules/QueueGranulesSpecPostProcessing.input.payload.json:9]: Agreed, will sort.

Comment (Jkovarik) [reply example/spec/parallel/queueGranules/QueueGranulesSpecPostProcessing.input.payload.json:9]: 682f31248 should sort it, waiting on a build to test.

File: example/spec/parallel/queueGranules/queueGranulesPostProcessingSpec.js
Comment (Jkovarik) [reply example/spec/parallel/queueGranules/queueGranulesPostProcessingSpec.js]: This one is a bit tricky.  The nature of the problem is the team when implementing the QueuedGranules feature didn't account for the lambda running *either* before SyncGranules (which has a custom/non-compliant granules object or *after* where it does, and createdAt is set. 

This test is testing that the queued granules workflow, with a database-compliant granule object/a populated createdAt value will succeed. 

How about: 6807c31a5

File: tasks/queue-granules/tests/index.js
Comment (Jkovarik) [reply tasks/queue-granules/tests/index.js]: 2922e830e

File: tasks/queue-granules/index.js
Comment (Jkovarik) [reply tasks/queue-granules/index.js:58]: I'm always a fan of more jsdocs. 

How's 6d1163994 look?

File: tasks/queue-granules/index.js
Comment (Jkovarik) [reply tasks/queue-granules/index.js:103]: > This got moved down, I think it should remain on L95


Hmm, I think the line numbers have shifted, but definitely before `const createdAt = Date.now();`

17f9f1d3b

File: tasks/queue-granules/tests/index.js
Comment (Jkovarik) [reply tasks/queue-granules/tests/index.js]: Hmm, that language is terrible 😓 .  I'd like to blame a merge, but it's new, so. 

It's testing that granules are added to the sqs queue, and they have  createdAt values.

Also this test is just junk, and reads like an interrupted thought.    Good catch.  👍🏻  

11dab5289
1e15d507e

File: tasks/queue-granules/tests/index.js
Comment (Jkovarik) [reply tasks/queue-granules/tests/index.js:393]: resolving in favor of 1e15d507e

File: tasks/queue-granules/tests/index.js
Comment (Jkovarik) [reply tasks/queue-granules/tests/index.js]: resolving in favor of 1e15d507e

File: tasks/queue-granules/tests/index.js
Comment (Jkovarik) [reply tasks/queue-granules/tests/index.js:404]: resolving in favor of 1e15d507e

File: tasks/queue-granules/tests/index.js
Comment (Jkovarik) [reply tasks/queue-granules/tests/index.js:809]: I don't *believe* that's needed as they're local scope/anon spies.

File: example/spec/parallel/queueGranules/queueGranulesPostProcessingSpec.js
Comment (Jkovarik) [reply example/spec/parallel/queueGranules/queueGranulesPostProcessingSpec.js:214]: I don't think so - the purpose of this integration test is to validate that granules being queued from queueGranules integrate with the step function event-driven write logic.  

However, we probably *should* verify the granule written is from the correct execution, as modifications to the queueGranules workflow could result in a test passing due to duplicate writes. 

dcdc4cab0

File: example/spec/parallel/queueGranules/QueueGranulesSpecPostProcessing.input.payload.json
Comment (Jkovarik) [reply example/spec/parallel/queueGranules/QueueGranulesSpecPostProcessing.input.payload.json:9]: 682f31248

**Comment by Jkovarik** [comment]
@flamingbear  thanks for the great review!  I think I've addressed the comments you left, take a look and let me know what you think.

File: tasks/queue-granules/tests/index.js
Comment (flamingbear) [reply tasks/queue-granules/tests/index.js]: much better.

File: example/spec/parallel/queueGranules/queueGranulesPostProcessingSpec.js
Comment (flamingbear) [reply example/spec/parallel/queueGranules/queueGranulesPostProcessingSpec.js:214]: That's a good point.  But why wouldn't you also want to check the createdAt hasn't been changed? I suppose it's not explicitly in the name of this test.

**Comment by flamingbear** [comment]
One question above and one suggested PR that I think you overlooked, or you hated and ignored https://github.com/nasa/cumulus/pull/2873

File: example/spec/parallel/queueGranules/queueGranulesPostProcessingSpec.js
Comment (Jkovarik) [reply example/spec/parallel/queueGranules/queueGranulesPostProcessingSpec.js:214]: It's covered better by units insofar as it relates to the component behavior, and generally I think we want to constrain that sort of evaluation to what's being integrated.     

Put another way, if the child workflow/step function invocation behavior/db write behavior is updated to enforce a modified createdAt time, we're not really testing either way with this integration test, we just want to be sure that *if* queuedGranules queues a granule with an existing created at value the framework records it in the child workflow.

**Comment by Jkovarik** [comment]
@flamingbear  I did indeed miss that comment/PR, looking.

**Comment by flamingbear** [comment]
> @flamingbear I did indeed miss that comment/PR, looking.
To be fair, it was just a comment before.

**Comment by Jkovarik** [comment]
@flamingbear merged you suggestion PR with changes as discussed.

**Comment by flamingbear** [comment]
Looks good, Wait until after Arch meeting to merge? Will approve either way.

**Review by flamingbear — APPROVED** [review]

**Comment by Jkovarik** [comment]
@flamingbear  yes, will wait for team discussion to merge.

**Pull request merged**
